slimta/envelope.py

```diff
@@ -157,34 +177,58 @@ def encode_7bit(self, encoder=None):
 
         """
         header_data, msg_data = self.flatten()
+        # header data is not supposed, in any case, to contain non-ascii chars
+        encoded_header_data = header_data.encode('ascii')
         try:
-            msg_data.encode('ascii')
-        except UnicodeDecodeError:
+            msg_data.decode('ascii')
+        except UnicodeError:
             if not encoder:
                 raise
-            self._encode_parts(header_data, msg_data, encoder)
+            self._encode_parts(encoded_header_data, msg_data, encoder)
 
-    def parse(self, data):
-        """Parses the given string or :class:`~email.message.Message` to
+    def parse_msg(self, msg):
+        """Parses the given :class:`~email.message.Message` to
         populate the :attr:`headers` and :attr:`message` attributes.
 
         :param data: The complete message, headers and message body.
-        :type data: :py:obj:`str` or :class:`~email.message.Message`
+        :type data: :class:`~email.message.Message`
 
         """
+
+        outfp = BytesIO()
+        BytesGenerator(outfp).flatten(msg, False)
+        data = outfp.getvalue()
+
+        if six.PY2:
+            data = data.encode()
+
+        self.parse(data)
+
+    def parse(self, data):
+        """Parses the given string to populate the :attr:`headers` and
+        :attr:`message` attributes.
+
+        :param data: The complete message, headers and message body.
+        :type data: :py:obj:`bytes`
+
+        """
+        check_argtype(data, bytes, 'data')
+
         if isinstance(data, Message):
-            outfp = cStringIO.StringIO()
+            outfp = cStringIO()
             Generator(outfp).flatten(data, False)
             data = outfp.getvalue()
```

Comment:
Oh, that's something I forgot, sorry, 

My general was more to no longer support Message as argument, what do you think ? 

In current master code, `Envelope.parse()` was only used with message as argument once, and I changed it already so that it usess `parse_msg()`.

It's an API breaking change, but maybe a py3 support worth a 3.0 (Are you adhering to [semver](http://semver.org/) btw ?)